### PR TITLE
[GPU] Reserve CPU resource for GPU inference

### PR DESCRIPTION
**Pull request by yangwang201911**
### Details:
 - Reserve CPU resource for GPU inference
 - One scenario: In AUTO CTPUT mode, reserve part of CPU resource for GPU inference and use remaining CPU resource for CPU inference.
 - Add test case about reserving CPU  from the remaining CPU resource reserved by GPU.

### Tickets:
 - CVS-99136

### CODE REVIEW

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp [hunk 1]
```diff
@@ -48,7 +48,20 @@ CompiledModel::CompiledModel(InferenceEngine::CNNNetwork &network,
         if (config.get_property(ov::internal::exclusive_async_requests)) {
             //exclusiveAsyncRequests essentially disables the streams (and hence should be checked first) => aligned with the CPU behavior
             return executorManager()->getExecutor("GPU");
-        }  else if (config.get_property(ov::num_streams) > 1) {
+        } else if (config.get_property(ov::num_streams) > 1) {
+            if (config.get_property(ov::hint::enable_cpu_pinning)) {
```
Comment (peterchen-intel) [inline src/plugins/intel_gpu/src/plugin/compiled_model.cpp:52]: ov::hint::enable_cpu_pinning  should be in plugin(RW) and CompiledModel (RO) supported list, implement the corresponding R/W code.

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp [hunk 2]
```diff
@@ -48,7 +48,20 @@ CompiledModel::CompiledModel(InferenceEngine::CNNNetwork &network,
...
+                auto executorConfig = ov::threading::IStreamsExecutor::Config{
+                    "Intel GPU plugin executor",
+                    0,
+                    0,
+                    InferenceEngine::IStreamsExecutor::ThreadBindingType::CORES,
+                    1,
+                    0,
+                    0,
+                    InferenceEngine::IStreamsExecutor::Config::PreferredCoreType::BIG,
+                    {{config.get_property(ov::num_streams), 1, 1, 0, 0}}};
```
Comment (peterchen-intel) [inline src/plugins/intel_gpu/src/plugin/compiled_model.cpp:62]: Update to all new API (merged)

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp (hunk 1 above)
Comment (yangwang201911) [reply src/plugins/intel_gpu/src/plugin/compiled_model.cpp:53]: Update.

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp (hunk 2 above)
Comment (yangwang201911) [reply src/plugins/intel_gpu/src/plugin/compiled_model.cpp:63]: update.

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp (hunk 1 above)
Comment (yangwang201911) [reply src/plugins/intel_gpu/src/plugin/compiled_model.cpp:53]: test case updated.

File: src/inference/src/dev/threading/cpu_streams_executor_internal.cpp [hunk 3]
```diff
@@ -127,23 +127,25 @@ void reserve_cpu_by_streams_info(const std::vector<std::vector<int>> _streams_in
...
-                _stream_processors[stream_pos[j]].push_back(_cpu_mapping_table[i][CPU_MAP_PROCESSOR_ID]);
-                _cpu_mapping_table[i][CPU_MAP_USED_FLAG] = _cpu_status;
-                if (static_cast<int>(_stream_processors[stream_pos[j]].size()) ==
-                    streams_table[j][THREADS_PER_STREAM]) {
-                    stream_pos[j]++;
-                    stream_num[j]++;
-                }
-                if (stream_num[j] >= streams_table[j][NUMBER_OF_STREAMS]) {
-                    stream_conditions[j].clear();
+        if (_cpu_mapping_table[i][CPU_MAP_USED_FLAG] == NOT_USED) {
```
Comment (yangwang201911) [inline src/inference/src/dev/threading/cpu_streams_executor_internal.cpp:130]: updated here from @sunxiaoxia2022

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp [hunk 4]
```diff
@@ -35,9 +35,27 @@ std::shared_ptr<ov::threading::ITaskExecutor> create_task_executor(const std::sh
...
+                                                        {{config.get_property(ov::num_streams), 1, 1, 0, 0}},
+                                                        true};
+            auto postConfig = ov::threading::IStreamsExecutor::Config::reserve_cpu_threads(executorConfig);
+            return std::make_shared<ov::threading::CPUStreamsExecutor>(postConfig);
+        }
         return std::make_shared<ov::threading::CPUStreamsExecutor>(
             ov::threading::IStreamsExecutor::Config{"Intel GPU plugin executor", config.get_property(ov::num_streams)});
+    } else {
+        return std::make_shared<ov::threading::CPUStreamsExecutor>(
+            ov::threading::IStreamsExecutor::Config{"Intel GPU plugin executor", 1});
```
Comment (yangwang201911) [inline src/plugins/intel_gpu/src/plugin/compiled_model.cpp:58]: @sunxiaoxia2022 need your help to review the changes here.

Comment (sunxiaoxia2022) [reply src/plugins/intel_gpu/src/plugin/compiled_model.cpp:58]: It is best to use macro ‘MAIN_CORE_PROC’ which is the second value in streams_info_table.

File: src/inference/src/dev/threading/cpu_streams_executor_internal.cpp (hunk 3 above)
Comment (sunxiaoxia2022) [reply src/inference/src/dev/threading/cpu_streams_executor_internal.cpp:130]: It is used in GPU reservation and responding test case is updated.

Comment (yangwang201911) [reply src/inference/src/dev/threading/cpu_streams_executor_internal.cpp:130]: Updated.

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp (hunk 4 above)
Comment (yangwang201911) [reply src/plugins/intel_gpu/src/plugin/compiled_model.cpp:58]: Updated.

**Review by sunxiaoxia2022 — APPROVED** [review]

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp [hunk 5]
```diff
@@ -35,9 +35,27 @@ std::shared_ptr<ov::threading::ITaskExecutor> create_task_executor(const std::sh
     if (config.get_property(ov::internal::exclusive_async_requests)) {
         //exclusive_async_requests essentially disables the streams (and hence should be checked first) => aligned with the CPU behavior
         return plugin->get_executor_manager()->get_executor("GPU");
-    } else {
+    } else if (config.get_property(ov::num_streams) > 1) {
```
Comment (wangleis) [inline src/plugins/intel_gpu/src/plugin/compiled_model.cpp:38]: CPU pinning will work for all number of streams. Please remove `if (config.get_property(ov::num_streams) > 1)`

**Review by wangleis — APPROVED** [review]

**Review by vladimir-paramuzov — APPROVED** [review]

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp (hunk 5 above)
Comment (yangwang201911) [reply src/plugins/intel_gpu/src/plugin/compiled_model.cpp]: Updated.

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp [hunk 6]
```diff
@@ -35,9 +35,27 @@ std::shared_ptr<ov::threading::ITaskExecutor> create_task_executor(const std::sh
...
+        if (config.get_property(ov::hint::enable_cpu_pinning)) {
+            auto executorConfig =
```
[the remark this answers is not in the document]
Comment (yangwang201911) [reply src/plugins/intel_gpu/src/plugin/compiled_model.cpp]: Updated.

File: src/plugins/intel_gpu/src/plugin/compiled_model.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (yangwang201911) [reply src/plugins/intel_gpu/src/plugin/compiled_model.cpp]: Updated.

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**